### PR TITLE
fix(abigen): safe ident underscore followed by numeric

### LINKED ISSUE #967: Abigen fails for identifier named _100pct
**Version**

```
├── ethers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   ├── ethers-addressbook v0.1.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   ├── ethers-contract v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   ├── ethers-contract-abigen v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-contract-derive v0.6.0 (proc-macro) (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-contract-abigen v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   ├── ethers-etherscan v0.2.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-solc v0.2.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   ├── ethers-middleware v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   ├── ethers-contract v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-etherscan v0.2.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   │   ├── ethers-signers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01)
│   │   │   ├── ethers-core v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   ├── ethers-providers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   ├── ethers-signers v0.6.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)
│   └── ethers-solc v0.2.0 (https://github.com/gakonst/ethers-rs#faba6e01) (*)                                                                                                                                          
```

**Platform**

```
Darwin Meets-Mac-mini.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T8101 arm64
```

**Description**

Abigen fails for identifier named `_100pct`. I am guessing this has to do with the particular case of `_` followed by a number. The transforming logic probably eliminates the `_` and starts with whatever is next, in this case, the digit which becomes an invalid identifier. `10_0pct_` was the ident generated for `_100pct`.

I tried this code:

```
{
  "inputs": [],
  "name": "_100pct",
  "outputs": [
    {
      "internalType": "uint256",
      "name": "",
      "type": "uint256"
    }
  ],
  "stateMutability": "view",
  "type": "function"
},
```

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes #967

## Solution
check if identifiers are valid and prefix with an underscore if they start with a numeric value

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

**Review by gakonst — APPROVED** [review]

**Comment by gakonst** [comment]
Merged to unblock the compilation in the  linked issue, but maybe we should fix inflector too?

**Comment by mattsse** [comment]
+1 on fixing this upstream, will look into the root cause.

File: ethers-contract/tests/abigen.rs [hunk 1]
```diff
@@ -374,6 +374,23 @@ fn can_handle_unique_underscore_functions() {
...
+            _100pct(string)
+        ]"#
+    );
+    let call = _100PctCall("message".to_string());
+
+    let provider = Arc::new(Provider::new(MockProvider::new()));
+    let contract = Test::new(Address::default(), Arc::clone(&provider));
+    // NOTE: this seems to be weird behaviour of `Inflector::to_snake_case` which turns "100pct" ->
+    // "10_0pct"
+    let _call = contract._10_0pct("hello".to_string());
```
Comment (meetmangukiya) [inline ethers-contract/tests/abigen.rs:391]: Do we know why it becomes `10_0pct` and not `100_pct`? @mattsse

Comment (gakonst) [reply ethers-contract/tests/abigen.rs:391]: Check the comment right above this code line! It's an issue in the Inflector crate, which we'll track and solve separately

**Pull request merged**